foundation/org.eclipse.persistence.core/src/org/eclipse/persistence/internal/queries/DatabaseQueryMechanism.java

```diff
@@ -490,6 +492,9 @@ public void insertObjectForWrite() {
             // Store the size of the modify row so we can determine if the user has added to the row in the insert.
             int modifyRowSize = modifyRow.size();
 
+            //391279: check the commitManager for scheduled events that match this insert and may need to be merged
+            commitManager.checkForDataModificationEventMerges(writeQuery, session);
```

Comment:
this does not look right to me. Isn't there a way to capture the id of the owner somewhere around the place where the 'postInsert' event is being created and in case 'id' can be extracted (possibly from some objectchangeset) just add it to the row instead of fixing the wrong event further down the road?

Reply:
I do agree. It is unusual to have to let the data modification event get create, only to go back and remove it later, but the problem I was hitting was the order of operations. 

The data modification events actually get added during 'postUpdate' for the UnidirectionalOneToMany side (In my examples, the Post class). During that time, we do indeed generate the 'id' for the other side (the Comment class) and it's that object that gets added to the event! However, we are still in 'postUpdate' for the UnidirectionalOneToMany mapping. We haven't even started the INSERT for the Comment object yet.

That code seems to be controlled here:

https://github.com/eclipse-ee4j/eclipselink/blob/3e83f76f7d4e7275d9d277a4375083ed0bd437d8/foundation/org.eclipse.persistence.core/src/org/eclipse/persistence/internal/sessions/CommitManager.java#L136-L142

See how we are looping through classes? One iteration will hit the 'postUpdate' for the Post class (which generates the data modification events), but a later iteration will actually INSERT the Comment objects. It's during THAT INSERT that I am checking for any data modification events, created in the past, that should be merged into the inserting query. 

The problem is that the data modification event is created so that we remember to UPDATE COMMENT after all the classes are processed (which unfortunately will be too late). We need the data modification event to be created so that we know an UPDATE WAS scheduled and we instead want to merge it into the current INSERT going on. 

I did try looking into this route, but the problem is the order of operations. Here is where the event gets create; you are correct, it's in 'postInsert':

https://github.com/eclipse-ee4j/eclipselink/blob/3e83f76f7d4e7275d9d277a4375083ed0bd437d8/foundation/org.eclipse.persistence.core/src/org/eclipse/persistence/mappings/OneToManyMapping.java#L993-L1003

Unfortunately, it's during postInsert for the UnidirectionalOneToMany side (In the example, POST); we haven't started INSERTs for the other side (In the example, COMMENT). 

The order goes like this:

First, we INSERT INTO POST. This triggers the postInsert(), which says "Make sure you register that this INSERT 

https://github.com/eclipse-ee4j/eclipselink/blob/3e83f76f7d4e7275d9d277a4375083ed0bd437d8/foundation/org.eclipse.persistence.core/src/org/eclipse/persistence/internal/sessions/CommitManager.java#L136-L142

First, we

Reply:
One thing maybe I should consider is moving my code to 'preInsert()' instead. I hadn't considered that before, but UnidirectionalOneToMany doesn't currently override 'preInsert()' which means I could just have the mapping listen for any INSERTS and check the commitManager at THAT time. Thoughts?

Reply:
what I've observed recently is that when I added backpointer mapping (leaving the ownership on the onetomany) for comparision, the difference was that instead of insert/update statements, just full inserts were used and the commit order computed by commitmanager seemed to be the same. I also don't think that the postInsert event was triggered in this case. That leads me to think that there should be a way to avoid doing updates (even for nullable=true cases eventually) Haven't had a chance to dig more into this due to different task unfortunately :-/

In any case, it may be good idea to introduce some property to control this.

Reply:
I tried adding a backpointer mapping. Unfortunately, I could not get around the fact that the mapping has no accessor and no actual field on the object. There are 101 places in EclipseLink where we use the field name (keys in maps for instance) that lead me to not being able to get around that limitation.